maven/JarAssembler.kt

```diff
@@ -34,35 +37,64 @@ class JarAssembler : Callable<Unit> {
     @Option(names = ["--jars"], split = ";")
     lateinit var jars: Array<File>
 
-    val entryNames = mutableSetOf<String>()
+    private val entryNames = mutableSetOf<String>()
+    val entries = HashMap<String, ByteArray>()
+
+    /**
+     * For path "a/b/c.java" inserts "a/" and "a/b/ into `entries`
+     */
+    private fun addDirectories(path: Path) {
+        for (i in path.nameCount-1 downTo 1) {
+            val subPath = path.subpath(0, i).toString() + "/"
+            entries[subPath] = ByteArray(0)
+        }
+    }
 
     override fun call() {
         ZipOutputStream(BufferedOutputStream(FileOutputStream(output_file))).use { out ->
             if (pomFile != null) {
-                val pomFileEntry = ZipEntry("META-INF/maven/${groupId}/${artifactId}/pom.xml")
-                out.putNextEntry(pomFileEntry)
-                out.write(pomFile!!.readBytes())
+                val pomPath = "META-INF/maven/${groupId}/${artifactId}/pom.xml"
+                entries[pomPath] = pomFile!!.readBytes()
+                addDirectories(Paths.get(pomPath))
             }
             for (jar in jars) {
                 ZipFile(jar).use { jarZip ->
                     jarZip.entries().asSequence().forEach { entry ->
                         if (entry.name.contains("META-INF")) {
+                            // pom.xml will be added by us
                             return@forEach
                         }
                         if (entryNames.contains(entry.name)) {
+                            throw RuntimeException("duplicate entry in the JAR: ${entry.name}")
+                        }
+                        if (entry.isDirectory) {
+                            // needed directories would be added by us
                             return@forEach
                         }
                         entryNames.add(entry.name)
                         BufferedInputStream(jarZip.getInputStream(entry)).use { inputStream ->
-                            val newEntry = ZipEntry(prefix + entry.name)
-                            out.putNextEntry(newEntry)
-                            inputStream.copyTo(out, 1024)
+                            val sourceFileBytes = inputStream.readBytes()
```

Comment:
can we extract this bit into a function that is named accordingly:
```
                            val sourceFileBytes = inputStream.readBytes()
                            val sourceFile = sourceFileBytes.toString(Charset.forName("UTF-8"))
                            var resultLocation = entry.name
                            // files in source JARs are moved according to their `package` statement
                            if (entry.name.endsWith(".java")) {
                                val sourceFileName = Paths.get(entry.name).fileName.toString()
                                val sourceFilePackage = javaPackageRegex.find(sourceFile)?.groups?.get(1)?.value ?: throw RuntimeException("could not obtain package of ${entry.name}")
                                resultLocation = "${sourceFilePackage.replace(".", "/")}/$sourceFileName"
                            }
                            entries[resultLocation] = sourceFileBytes
                            addDirectories(Paths.get(resultLocation))
```

Reply:
@lolski addressed in https://github.com/vaticle/bazel-distribution/pull/305/commits/a843c9a2834b9693197d6e45ec799274a61284b2